### PR TITLE
feat: do not attempt to vote on requests more than once

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/684

## Changes

* Update the queries for `get_pending_deposit_requests` and `get_pending_withdrawal_requests` to ignore requests that the signer has voted on already.
* Update the in memory implementations of both functions.
* Add a helper query to maintain the same behavior in the `load_latest_deposit_requests_persists_requests_from_past` test.

## Testing Information

This PR adds integration tests that check the new behavior of the updated queries.

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

**Review by cylewitruk — APPROVED** [review]
Looks good :+1:

**Review by matteojug — APPROVED** [review]
The only downside of this is that if a signer doesn't get the vote message (e.g., it's down), it will not get it anymore. As long as >=threshold signers get it we should be fine though.

File: signer/src/storage/in_memory.rs [hunk 1]
```diff
@@ -187,6 +186,104 @@ impl Store {
...
+        let context_window_end_block =
+            futures::stream::unfold(bitcoin_chain_tip.block_hash, |block_hash| async move {
+                self.bitcoin_blocks
+                    .get(&block_hash)
+                    .map(|block| (block.clone(), block.parent_hash))
+            })
+            .skip(context_window as usize)
+            .boxed()
+            .next()
+            .await;
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/in_memory.rs:256]: We don't. When I was making these changes I briefly tried to think of a way to remove it and nothing came immediately to mind. But I just thought of a good way to do it, so I'll do that.

File: signer/src/testing/storage/postgres.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+//!
+
+use crate::error::Error;
+use crate::storage::model;
+use crate::storage::postgres::PgStore;
+
+impl PgStore {
+    /// Get all deposit requests that have been confirmed within the
+    /// context window.
+    pub async fn get_deposit_requests(
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/testing/storage/postgres.rs:11]: So far it doesn't seem so useful, but I'm opposed to adding it. I just didn't want to do the work of adding it (add another test, proper documentation, and probably some simplification).

File: signer/src/storage/in_memory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/in_memory.rs]: simplifyed in https://github.com/stacks-network/sbtc/pull/968/commits/dbe2bde80ab096e9e8a2b561657678d17bf10b82 and https://github.com/stacks-network/sbtc/pull/968/commits/e806d31e3c43684e92d82a4601841b7d2e875b7c

File: signer/tests/integration/postgres.rs [hunk 3]
```diff
@@ -355,22 +359,132 @@ async fn should_return_the_same_pending_deposit_requests_as_in_memory_store() {
...
 
-    let mut pending_deposit_requests = in_memory_store
-        .get_pending_deposit_requests(&chain_tip, context_window)
+    for signer_public_key in signer_set.iter() {
+        let mut pending_deposit_requests = in_memory_store
+            .get_pending_deposit_requests(&chain_tip, context_window, signer_public_key)
+            .await
+            .expect("failed to get pending deposit requests");
+
+        pending_deposit_requests.sort();
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/tests/integration/postgres.rs:368]: Done in https://github.com/stacks-network/sbtc/pull/968/commits/1fcc6c4fc359923a3132befd41093c1e4da28c57

File: signer/tests/integration/postgres.rs [hunk 4]
```diff
@@ -355,22 +359,132 @@ async fn should_return_the_same_pending_deposit_requests_as_in_memory_store() {
...
+    let mut rng = rand::rngs::StdRng::seed_from_u64(43);
+
+    let amounts = DepositAmounts { amount: 123456, max_fee: 12345 };
+    let signers = TestSignerSet::new(&mut rng);
+    let setup = TestSweepSetup2::new_setup(signers, faucet, &[amounts]);
+
+    backfill_bitcoin_blocks(&db, rpc, &setup.deposit_block_hash).await;
+    let chain_tip = db.get_bitcoin_canonical_chain_tip().await.unwrap().unwrap();
+
+    // There aren't any deposit requests in the database.
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/tests/integration/postgres.rs]: Done in https://github.com/stacks-network/sbtc/pull/968/commits/1fcc6c4fc359923a3132befd41093c1e4da28c57

File: signer/src/testing/storage/postgres.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+                UNION ALL
+
+                -- Recursive member: Fetch the parent block using the last block's parent_hash
+                SELECT parent.block_hash, parent.block_height, parent.parent_hash,
+                       parent.created_at, last.depth + 1
+                FROM sbtc_signer.bitcoin_blocks parent
+                JOIN context_window last ON parent.block_hash = last.parent_hash
+                WHERE last.depth < $2
+            ),
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/testing/storage/postgres.rs]: Updated in https://github.com/stacks-network/sbtc/pull/968/commits/c6f54eedf157c6572b9550498352ae02e2f13060

File: signer/src/testing/storage/postgres.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
+//! A module with helper queryy functions.
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/testing/storage/postgres.rs]: Fixed in https://github.com/stacks-network/sbtc/pull/968/commits/c6f54eedf157c6572b9550498352ae02e2f13060

File: signer/src/testing/storage/postgres.rs [hunk 7]
```diff
@@ -15,41 +15,20 @@ impl PgStore {
...
+              , dr.reclaim_script
+              , dr.recipient
+              , dr.amount
+              , dr.max_fee
+              , dr.lock_time
+              , dr.signers_public_key
+              , dr.sender_script_pub_keys
+            FROM sbtc_signer.bitcoin_blockchain_of($1, $2)
+            JOIN sbtc_signer.bitcoin_transactions USING (block_hash)
+            JOIN sbtc_signer.deposit_requests AS dr USING (txid)
```
Comment (matteojug) [inline signer/src/testing/storage/postgres.rs:31]: will this work fine if we have multiple deposits in a single bitcoin block?

Comment (djordon) [reply signer/src/testing/storage/postgres.rs:31]: It should, since each of them will have a unique transaction ID.

Comment (matteojug) [reply signer/src/testing/storage/postgres.rs:31]: Nvm, my bad!

**Pull request merged**